app/models/binary/explore/ExploreRemoteLayerService.scala

```diff
@@ -66,11 +70,62 @@ class ExploreRemoteLayerService @Inject()() extends FoxImplicits with LazyLoggin
     }
   }
 
-  private def commonVoxelSize(voxelSizes: List[Vec3Double])(implicit ec: ExecutionContext): Fox[Vec3Double] =
+  private def magFromVoxelSize(minVoxelSize: Vec3Double, voxelSize: Vec3Double)(
+      implicit ec: ExecutionContext): Fox[Vec3Int] = {
+    def isPowerOfTwo(x: Int): Boolean =
+      x != 0 && (x & (x - 1)) == 0
+
+    val mag = (voxelSize / minVoxelSize).round.toVec3Int
+    for {
+      _ <- bool2Fox(isPowerOfTwo(mag.x) && isPowerOfTwo(mag.x) && isPowerOfTwo(mag.x)) ?~> s"invalid mag: $mag. Must all be powers of two"
+    } yield mag
+  }
+
+  private def checkForDuplicateMags(magGroup: List[Vec3Int])(implicit ec: ExecutionContext): Fox[Unit] =
+    for {
+      _ <- bool2Fox(magGroup.length == 1) ?~> s"detected mags are not unique, found $magGroup"
+    } yield ()
+
+  private def rescaleLayersByCommonVoxelSize(layersWithVoxelSizes: List[(DataLayer, Vec3Double)])(
+      implicit ec: ExecutionContext): Fox[(List[DataLayer], Vec3Double)] = {
+    val allVoxelSizes = layersWithVoxelSizes
+      .flatMap(layerWithVoxelSize => {
+        val layer = layerWithVoxelSize._1
+        val voxelSize = layerWithVoxelSize._2
+
+        layer.resolutions.map(resolution => voxelSize * resolution.toVec3Double)
+      })
+      .toSet
+    val minVoxelSizeOpt = Try(allVoxelSizes.minBy(_.toTuple)).toOption
+
     for {
-      head <- voxelSizes.headOption.toFox
-      _ <- bool2Fox(voxelSizes.forall(_ == head)) ?~> s"voxel sizes for layers are not uniform, got $voxelSizes"
-    } yield head
+      minVoxelSize <- option2Fox(minVoxelSizeOpt)
+      allMags <- Fox.combined(allVoxelSizes.map(magFromVoxelSize(minVoxelSize, _)).toList) ?~> s"voxel sizes for layers are not uniform, got ${layersWithVoxelSizes
+        .map(_._2)}"
+      groupedMags = allMags.groupBy(_.maxDim)
+      _ <- Fox.combined(groupedMags.values.map(checkForDuplicateMags).toList)
+      rescaledLayers = layersWithVoxelSizes.map(layerWithVoxelSize => {
+        val layer = layerWithVoxelSize._1
+        val layerVoxelSize = layerWithVoxelSize._2
+        val magFactors = (layerVoxelSize / minVoxelSize).toVec3Int
+        layer match {
+          case l: ZarrDataLayer =>
+            l.copy(mags = l.mags.map(mag => mag.copy(mag = mag.mag * magFactors)),
+                   boundingBox = l.boundingBox * magFactors)
+          case l: ZarrSegmentationLayer =>
+            l.copy(mags = l.mags.map(mag => mag.copy(mag = mag.mag * magFactors)),
+                   boundingBox = l.boundingBox * magFactors)
+          case l: N5DataLayer =>
+            l.copy(mags = l.mags.map(mag => mag.copy(mag = mag.mag * magFactors)),
+                   boundingBox = l.boundingBox * magFactors)
+          case l: N5SegmentationLayer =>
+            l.copy(mags = l.mags.map(mag => mag.copy(mag = mag.mag * magFactors)),
+                   boundingBox = l.boundingBox * magFactors)
```

Comment:
As I understand it this could be simplified by using | to match multiple cases. Although I think this would require a new trait common to datasets (since DataSet/DataSetLike does not have e.g. mags)

Reply:
@fm3 Do you have a good idea how this could be improved? Using a common trait is difficult, because the code relies on the `copy` method, which is only defined for case classes.